File: frontend/src/lib/portfolioOverview.js
import rules from './portfolioRules.json';
import { complianceNavigation } from './complianceNavigation';

// A projection of the existing management populations, not another issue engine.
// Keep the legacy combined metric intact for other consumers.
export function portfolioPopulations(model) {
  return {
    critical_high_issues: model.metrics.critical_high_open.filter(r => r.kind !== 'risks'),
    significant_risks: model.significantRisks
  };
}
export function portfolioOrder(a, b) {
  for (const key of rules.priorityOrder) {
    const difference = b[key] - a[key];
    if (difference) return difference;
  }
  return a.name.localeCompare(b.name, undefined, {
    sensitivity: 'base'
  }) || a.client_id.localeCompare(b.client_id);
}
export function portfolioFrameworks(clientId, baseline, requirements) {
  return complianceNavigation(clientId, baseline, requirements).map(({
    key,
    label,
    to
  }) => ({
    key,
    label,
    to
  }));
}
export function meaningfulActivity(log, now = new Date()) {
  const definition = rules.activity.find(r => [r.kind, r.kind === 'policy' ? 'policies' : r.kind + 's'].includes(log.entity_type) && r.actions.includes(log.action));
  const time = Date.parse(log.at);
  if (!definition || !Number.isFinite(time) || time > now.getTime()) return null;
  return {
    at: log.at,
    label: definition.labels?.[log.action] || log.action
  };
}
export function latestPortfolioActivity(logs, now = new Date()) {
  const latest = new Map();
  for (const log of logs) {
    const activity = meaningfulActivity(log, now),
      previous = latest.get(log.client_id);
    if (activity && (!previous || Date.parse(activity.at) > Date.parse(previous.at))) latest.set(log.client_id, activity);
  }
  return latest;
}

File: backend/portfolio_overview.py
"""Read-only portfolio projections over authoritative management populations."""
import json
from datetime import datetime
from framework_governance import ROOT, FRAMEWORKS

RULES = json.loads((ROOT / 'portfolioRules.json').read_text())


def populations(model):
    return {'critical_high_issues': [r for r in model['metrics']['critical_high_open'] if r['kind'] != 'risks'],
            'significant_risks': model['significantRisks']}


def frameworks(client, requirements):
    selected = {r.get('baseline_key') for r in requirements
                if r.get('client_id') == client['client_id'] and r.get('baseline_response') == 'applies'}
    if not client.get('onboarding_baseline', {}).get('completed'):
        return []
    return [{'key': f['key'], 'label': f['label'], 'to': '/compliance/' + f['key']}
            for f in FRAMEWORKS if f['key'] in selected]


def attention_order(row):
    return (*(-row[key] for key in RULES['priorityOrder']), row['name'].casefold(), row['client_id'])


async def latest_activity(db, client_ids, now):
    # One aggregate for authorized clients. Return one narrow lifecycle event per
    # client rather than downloading the entire audit history or generic updates.
    clauses = [{'entity_type': {'$in': [r['kind'], 'policies' if r['kind'] == 'policy' else r['kind'] + 's']},
                'action': {'$in': r['actions']}} for r in RULES['activity']]
    logs = await db.audit_logs.aggregate([
        {'$match': {'client_id': {'$in': client_ids}, '$or': clauses,
                    'at': {'$type': 'string', '$lte': now}}},
        {'$sort': {'at': -1}},
        {'$group': {'_id': '$client_id', 'at': {'$first': '$at'},
                    'action': {'$first': '$action'}, 'entity_type': {'$first': '$entity_type'}}},
    ]).to_list(None)
    latest = {}
    for log in logs:
        try:
            at = datetime.fromisoformat(log['at'].replace('Z', '+00:00'))
            if at.tzinfo is None or at > datetime.fromisoformat(now.replace('Z', '+00:00')):
                continue
        except (ValueError, TypeError):
            continue
        definition = next(r for r in RULES['activity'] if log['action'] in r['actions']
                          and log['entity_type'] in (r['kind'], 'policies' if r['kind'] == 'policy' else r['kind']+'s'))
        latest[log['_id']] = {'at': log['at'], 'label': definition.get('labels', {}).get(log['action'], log['action'])}
    return latest
